### PR TITLE
Plugin cors

### PR DESCRIPTION
**Pull request by 2456868764**
### Ⅰ. Describe what this PR did
Add CORS plugin refer to issue  #336

### CODE REVIEW

File: plugins/wasm-go/extensions/cors/README.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  - '*'
+allow_methods:
+  - '*'  
+allow_headers:
+  - '*'
+expose_headers:
+  - '*'
+allow_credentials: false
+max_age: 7200
+
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/cors/README.md:35]: I don't think we need an extra blank line at the end of these configurations.

**Review by CH3CHO — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+const (
+	defaultMatchAll     = "*"
+	defaultAllowMethods = "GET, PUT, POST, DELETE, PATCH, OPTIONS"
+	defaultAllowHeaders = "DNT,Keep-Alive,User-Agent,X-Requested-With," +
+		"If-Modified-Since,Cache-Control,Content-Type,Authorization"
+	defaultAllowCredentials = false
+	defaultMaxAge           = 7200
+
+	defaultBadHttpStatus       = 403
+	defaultBadHttpResponseBody = "Invalid CORS request"
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+func (c *CorsConfig) GetVersion() string {
+	return "1.0.0"
+}
+
+func (c *CorsConfig) FillDefaultValues() {
+	if len(c.allowOrigins) == 0 && len(c.allowOriginPatterns) == 0 && c.allowCredentials == false {
+		c.allowOrigins = []string{defaultMatchAll}
+	}
+	if len(c.allowHeaders) == 0 {
+		c.allowHeaders = []string{defaultMatchAll}
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+	"strings"
+)
+
+const (
+	defaultMatchAll     = "*"
+	defaultAllowMethods = "GET, PUT, POST, DELETE, PATCH, OPTIONS"
+	defaultAllowHeaders = "DNT,Keep-Alive,User-Agent,X-Requested-With," +
+		"If-Modified-Since,Cache-Control,Content-Type,Authorization"
+	defaultAllowCredentials = false
+	defaultMaxAge           = 7200
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

File: plugins/wasm-go/extensions/cors/main.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	log.Debugf("corsConfig:%+v", corsConfig)
+	return nil
+}
+
+func onHttpRequestHeaders(ctx wrapper.HttpContext, corsConfig config.CorsConfig, log wrapper.Log) types.Action {
+	log.Debug("onHttpRequestHeaders()")
+	requestUrl, _ := proxywasm.GetHttpRequestHeader(":path")
+	method, _ := proxywasm.GetHttpRequestHeader(":method")
+	host := ctx.Host()
+	schema := ctx.Scheme()
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/main.go]: OK

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+
+	defaultBadHttpStatus       = 403
+	defaultBadHttpResponseBody = "Invalid CORS request"
+
+	protocalHttpName  = "http"
+	protocalHttpPort  = "80"
+	protocalHttpsName = "https"
+	protocalHttpsPort = "443"
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+
+	HeaderPluginDebug = "X-Cors-Version"
+	HeaderPluginTrace = "X-Cors-Trace"
+	HeaderOrigin      = "Origin"
+	HTTPMethodOptions = "OPTIONS"
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

File: plugins/wasm-go/extensions/cors/README.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+|-----------------------|-----------------|-------|------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| allow_origins         | array of string | 选填    | *                                                                                                          | 允许跨域访问的 Origin，格式为 scheme://host:port，示例如 http://example.com:8081。当 allow_credentials 为 false 时，可以使用 * 来表示允许所有 Origin 通过                                                     |
+| allow_origin_patterns | array of string | 选填    | -                                                                                                          | 允许跨域访问的 Origin 模式匹配， 用 * 匹配域名或者端口， <br/>比如 http://*.example.com -- 匹配域名， http://*.example.com:[8080,9090] -- 匹配域名和指定端口， http://*.example.com:[*] -- 匹配域名和所有端口。单独 * 表示匹配所有域名和端口 |
```
[the remark this answers is not in the document]
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/README.md]: OK

File: plugins/wasm-go/extensions/cors/README.md (hunk 1 above)
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/README.md]: OK

**Comment by 2456868764** [comment]
i'v updated , please review again

**Review by CH3CHO — APPROVED** [review]
LFTM

**Review by CH3CHO — APPROVED** [review]
LFTM

**Review by CH3CHO — APPROVED** [review]
LGTM

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+	origin = strings.TrimSpace(origin)
+	if len(origin) == 0 {
+		return "", false
+	}
+
+	matchOrigin := strings.ToLower(origin)
+	// Check exact match
+	for _, allowOrigin := range c.allowOrigins {
+		if allowOrigin == defaultMatchAll {
+			return defaultMatchAll, true
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/cors/config/cors_config.go:339]: It's better to return origin, which is the  raw origin header in the request, methods and headers need to be treated similarly, returning wildcards will not work in some browsers.

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/cors/config/cors_config.go (hunk 9 above)
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

**Comment by 2456868764** [comment]
i'v updated, please review again.

File: plugins/wasm-go/extensions/cors/config/cors_config.go [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+
+func (c *CorsConfig) checkMethods(requestMethod string) (string, bool) {
+	if len(requestMethod) == 0 {
+		return "", false
+	}
+
+	// Find method existed in allowMethods configuration
+	for _, method := range c.allowMethods {
+		if method == defaultMatchAll {
+			return defaultMatchAll, true
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/cors/config/cors_config.go:396]: It's better to return  all raw methods like "GET,POST,PUT,PATCH..."

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/cors/config/cors_config.go (hunk 10 above)
Comment (2456868764) [reply plugins/wasm-go/extensions/cors/config/cors_config.go]: OK

**Review by johnlanni — APPROVED** [review]
LGTM

**Comment by 2456868764** [comment]
i'v updated, please review again.

**Pull request merged**